crates/astria-conductor/src/executor/mod.rs

```diff
@@ -369,52 +430,67 @@ impl Executor {
         block: ReconstructedBlock,
     ) -> eyre::Result<()> {
         let executable_block = ExecutableBlock::from_reconstructed(block);
-        let expected_height = self.state.borrow().next_firm_sequencer_height();
+        let expected_height = self.state.next_expected_firm_sequencer_height();
+        let block_height = executable_block.height;
         ensure!(
-            executable_block.height == expected_height,
-            "expected block at sequencer height {expected_height}, but got {}",
-            executable_block.height,
+            block_height == expected_height,
+            "expected block at sequencer height {expected_height}, but got {block_height}",
         );
 
-        let update_type = if let Some(block) = self
-            .blocks_pending_finalization
-            .remove(&executable_block.hash)
-        {
-            Update::OnlyFirm(block)
-        } else {
-            let parent_block_hash = self.state.borrow().firm_parent_hash();
-            let executed_block = self
-                .execute_block(client.clone(), parent_block_hash, executable_block)
-                .await
-                .wrap_err("failed to execute block")?;
-            does_block_response_fulfill_contract(
-                ExecutionKind::Firm,
-                &self.state.borrow(),
-                &executed_block,
+        let genesis_height = self.state.sequencer_genesis_block_height();
+        let Some(block_number) =
+            state::map_sequencer_height_to_rollup_height(genesis_height, block_height)
+        else {
+            bail!(
+                "failed to map block height rollup number. This means the operation
+                `sequencer_height - sequencer_genesis_height` underflowed or was not a valid
+                cometbft height. Sequencer height: `{block_height}`, sequencer genesis height: \
+                 `{genesis_height}`",
             )
-            .wrap_err("execution API server violated contract")?;
-            Update::ToSame(executed_block)
         };
+
+        let update_type =
+            if let Some(block) = self.blocks_pending_finalization.remove(&block_number) {
+                info!(
+                    block_number,
+                    "found pending block; updating state but not not re-executing it"
+                );
+                Update::OnlyFirm(block)
+            } else {
+                info!(
+                    block_number,
+                    "did not find pending block; executing firm block"
+                );
+                // The parent hash of the next block is the hash of the block at the current head.
+                let parent_hash = self.state.firm_hash();
```

Comment:
The entire logic is faulty: if `sort > firm` and there is no soft block number, then that's a hard error and the block must not be executed.

As you note, this code is not new but can be traced back to the original implementation last year. I am going to address this in a follow-up PR.

(Actually, this begs the question: why batch-request all these blocks in the first place? We could just as well request the parent-block from the rollup as we go instead of maintaining a cache. I am going to leave the implementation as it is for now, but just food for thought.)

Reply:
https://github.com/astriaorg/astria/issues/1011